src/software/thunderscope/binary_context_managers/full_system.py

```diff
@@ -0,0 +1,174 @@
+import os
+import logging
+import time
+import threading
+
+from subprocess import Popen
+from software.python_bindings import *
+from proto.import_all_protos import *
+from software.py_constants import *
+from software.thunderscope.binary_context_managers.binary_context_managers import (
+    is_cmd_running,
+)
+
+
+class FullSystem(object):
+
+    """ Full System Binary Context Manager """
+
+    def __init__(
+        self,
+        full_system_runtime_dir: os.PathLike = None,
+        debug_full_system: bool = False,
+        friendly_colour_yellow: bool = False,
+        should_restart_on_crash: bool = True,
+    ) -> None:
+        """Run FullSystem
+
+        :param full_system_runtime_dir: The directory to run the blue full_system in
+        :param debug_full_system: Whether to run the full_system in debug mode
+
+        """
+        self.full_system_runtime_dir = full_system_runtime_dir
+        self.debug_full_system = debug_full_system
+        self.friendly_colour_yellow = friendly_colour_yellow
+        self.full_system_proc = None
+        self.should_restart_on_crash = should_restart_on_crash
+
+        self.thread = threading.Thread(target=self.__restart__)
+
+    def __enter__(self) -> self:
```

Comment:
When I run Thunderscope, I get the following error:
```python
full_system.py", line 40, in FullSystem
    def __enter__(self) -> self:
NameError: name 'self' is not defined
```

You can refer to this comment for some more detail on adding typehinting here https://github.com/UBC-Thunderbots/Software/pull/3025#discussion_r1363697658

Reply:
After fixing this, pls try running `thunderscope_main` and some simulated pytests to verify that this is fixed